source/agora/common/Amount.d

```diff
@@ -0,0 +1,239 @@
+    public static immutable Amount MaxUnitSupply =
+        Amount(UnitPerCoin.value * MaxCoinSupply, true);
+
+    /// Internal data storage
+    private ulong value;
+
+    /// Support for hashing
+    public void computeHash(scope HashDg dg) const nothrow @nogc @safe
+    {
+        hashPart(this.value, dg);
+    }
+
+    /// Support for serialization
+    public void serialize(scope SerializeDg dg) const nothrow @safe
+    {
+        serializePart(this.value, dg);
+    }
+
+    /// Support for deserialization
+    public void deserialize(scope DeserializeDg dg) nothrow @safe
+    {
+        deserializePart(this.value, dg);
+    }
+
+    /// Pretty-print this value
+    public void toString (SinkT dg) const @safe
+    {
+        import std.format;
+        formattedWrite(dg, "%d", this.value);
+    }
+
+    /// Also support for Vibe.d serialization to JSON
+    public string toString () const @safe
+    {
+        string ret;
+        scope SinkT dg = (scope v) { ret ~= v; };
+        this.toString(dg);
+        return ret;
+    }
+
+    private alias SinkT = void delegate(scope const(char)[] v) @safe;
+
+    /// Support for Vibe.d deserialization
+    public static Amount fromString (scope const(char)[] str) pure @safe
+    {
+        import std.conv : to;
+        immutable ul = str.to!ulong;
+        if (!Amount.isInRange(ul))
+            throw new Exception("Invalid input value to Amount");
+        return Amount(ul, true);
+    }
+
+
+    nothrow pure @nogc @safe:
+
+    /***************************************************************************
+
+        Construct an instance of an `Amount`
+
+        Params:
+            units = The monetary amount this new instance represent.
+                    The unit is the smaller unit of currency (see `UnitPerCoin`)
+
+    ***************************************************************************/
+
+    public this (ulong units)
+    {
+        assert(Amount.isInRange(units));
+        this.value = units;
+    }
+
+    /// Make a value without checking bounds
+    /// Used to initialize the bounds themselves, and to make invalid values
+    /// in unittests
+    private this (ulong units, bool dummy) { this.value = units; }
+    /// Ditto
+    version (unittest)
+    public static Amount invalid (ulong units) { return Amount(units, true); }
+
+    /***************************************************************************
+
+        Returns:
+            `true` if `value` is in the range `[0; Amount.MaxUnitSupply]`
+            and can be safely used to make an `Amount`
+
+    ***************************************************************************/
+
+    pragma(inline, true)
+    public static bool isInRange (ulong value)
+    {
+        return  value <= MaxUnitSupply.value;
+    }
+
+    /// `isInRange` but as member function
+    pragma(inline, true)
+    public bool isValid () const
+    {
+        return isInRange(this.value);
+    }
+
```

Comment:
Two blank lines? What are you compensating for?

Reply:
https://www.youtube.com/watch?v=Dkk9gvTmCXY